### ISSUE TITLE
Details of two flavors of asserts

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#18024]
Follow-up to #16810 and #12466.
Related to #13603.

We're thinking that we would like to have two different variations on `assert`.
 * one would be always on (even with `--fast`)
 * one would be toggle-able with a `chpl` flag and normally be disabled with `--fast`

The rationale for having some `assert` be able to be always on is that if a program is encountering a bug that would lead to undefined behavior it would be better to fail fast rather than possibly running in to any other horrible consequence that could come from the undefined behavior. Nonetheless, an always-on assert is not a substitute for using `throw` or other reasonable error handling techniques in libraries.

Rust has this idea and calls the always-on one `assert` and the disabled-in-release `debug_assert`.

The goal of this issue is to discuss and try to sort out the details.

Some questions:
 * Should the default `assert` behavior be toggled off with `--fast` or no?
 * Should we have  different methods like `assert`  / `releaseAssert` / `debugAssert`, or should we have an optional argument to `assert` that indicates which of these it is? What are the exact names of these functions / arguments?

**Comment by bradcray** [comment]
Based on today's discussion, I here's the mental strawperson I ended up with:

```chapel
config param defaultPermitOpt=true,
                       defaultHaltOnFaile=true;
proc assert(test: bool, message: string = "", param permitOpt=defaultPermitOpt, haltOnFail=defaultHaltOnFail) { ... }
```

where:
* `permitOpt` says whether or not a flag like `--no-assert-checks` or `--no-checks` or `--fast` would permit it to be optimized away by the compiler
* `haltOnFail` would say whether or not failing the assertion should halt the program or print the message and continue

This second toggle (and therefore third behavior) was mentioned as desirable by @jhh67 in the meeting, which was an intriguing possibility that hadn't occurred to me before.

I think the main reluctance I have about the above is that the `config param`s are global to the program, so suggest that all libraries using asserts are abiding by the same rules.  However:
* library writers could always use the discipline of passing the arguments explicitly for their calls, or even introducing their own configs to set the defaults for those explicit actuals
* it was pointed out that in a logging-based system (frequently requested by Chapel users, so a likely future for us), the asserts might be connected to a specific logging channel/file/stream, in which case the defaults might be associated with that log or stream rather than using the globals here.

(and then if we were to provide the ability to unify compile-time asserts with these, I might provide an overload:

```chapel
proc assert(param test: bool, param message: string = "", param permitOpt=defaultPermitOpt, param haltOnFail=defaultHaltOnFail, param compileTime=false) { ... }
```

where the `haltOnFail` case is particularly interesting to me here because I find our compiler often asserts things, fails, and then it turns out if I remove the assert everything would've worked fine (i.e., someone was being overly paranoid of thought they were being overly clever).  So if I were able to recompile, asking it to blast past the assert without code changes / recompiling to see what happens, that would be a nice productivity boost.

**Comment by mppf** [comment]
I guess my concern with adding a bunch of optional arguments is that `assert` is arguably already a convenience function that is short for `if ! something { halt("fail"); }`. Meaning, rather than have `assert` support all of the possibilities, I think it should just do one thing that commonly occurs. I also find say `assert(x)` vs `debugAssert(x)` less mental overhead than `assert(permitOpt=true, x)`.

**Comment by jhh67** [comment]
After giving this more thought I think we should keep it simple for now. Eventually Chapel will have to do logging for real and that's probably the right time to tackle asserts. If we make it too complicated developers are less likely to use it or will use it incorrectly. I like the idea of an assert that is not disabled by --fast and one that is. Developers can understand this and the barrier for use is pretty low. Anything more complicated can likely done with exceptions.

**Comment by mppf** [comment]
@jhh67 - do you have a preference for what the names are (and in particular, if one of the behaviors is what you get with `assert()` ?)

**Comment by vasslitvinov** [comment]
I like the idea of `assert` taking an optional param argument indicating whether or not to check it upon --fast. I think it is Brad's `permitOpt`, although I'd like a clearer name (which is not coming to me at the moment). I suggest the default being "yes, check it upon --fast".

Another thing we ought to do is to forego evaluating the assert's argument when the assert is not checked. Ex. do not compute f() in assert(f()). If we do this while having the param argument to assert, this will require some AST handling in the parser to implement it correctly, akin to short-circuiting && and ||. This will turn `assert` into a built-in function, as opposed to something to opt in by use-ing or importing a module. Still, I suggest that it is worth it.

**Comment by vasslitvinov** [comment]
P.S. I am also open to go the Rust route with assert and debugAssert. Which BTW we should take for our compiler code as well. My desire to forego the evaluation of debugAssert's argument expression is still there.

**Comment by mppf** [comment]
> If we do this while having the param argument to assert, this will require some AST handling in the parser to implement it correctly, akin to short-circuiting && and ||. This will turn `assert` into a built-in function, as opposed to something to opt in by use-ing or importing a module. Still, I suggest that it is worth it.

This is a bit of a side issue here - but I don't agree - in particular I think it's reasonable for the compiler to have this particular way of compiling an `assert` function that is defined in the modules. It's true that then the assert function becomes "part of the language" but it still seems reasonable to me to have it be defined in module code.

**Comment by bradcray** [comment]
To make sure I'm understanding: It sounds to me as though @vasslitvinov and @mppf are wanting the same thing (assert doesn't evaluate its argument when disabled) and only disagree on whether there is code in the module to help implement it or not, but then agree again that it's a special function that the user couldn't write themselves since it requires compiler help.  Is that right?

**Comment by vasslitvinov** [comment]
I have a preference for `assert` and `debugAssert` to be (a) always-available rather than (b) require the use/import of a module. However this is not a strong preference.

My concern is that it is would require more compiler code additions to implement a not-always-evaluating assert function of the (b) kind than of the (a) kind.

**Comment by mppf** [comment]
> I have a preference for `assert` and `debugAssert` to be (a) always-available rather than (b) require the use/import of a module. However this is not a strong preference.

Right, I think we ended up already settling on these always being available (we decided to put them in an use'd-by-default module). However they can be invoked by e.g. `Errors.assert` in case one wants to disambiguate.

> To make sure I'm understanding: It sounds to me as though @vasslitvinov and @mppf are wanting the same thing (assert doesn't evaluate its argument when disabled) and only disagree on whether there is code in the module to help implement it or not, but then agree again that it's a special function that the user couldn't write themselves since it requires compiler help. Is that right?

I think that's right.

**Comment by jhh67** [comment]
> @jhh67 - do you have a preference for what the names are (and in particular, if one of the behaviors is what you get with `assert()` ?)

No, I don't have a preference. I'm bad at naming things.

**Comment by mppf** [comment]
We had some discussion of this issue and the leading proposals were:

* just add a global param to indicate if checks are on / in debug mode and to have an assert not enabled in release mode you would write e.g. `if checksAreOn then assert(...)` or `if debug then assert(...)`
* add an assert statement to the language (due to concern about a function that does not evaluate its arguments, even if only in release mode)
* don't provide `assert` but provide `releaseAssert` and `debugAssert` instead. We could have `debugAssert` evaluate its argument unless the compiler can prove it to be side effect free.

Of these, the `if checksAreOn then assert(...)` approach had the most support.

**Comment by bradcray** [comment]
I find the assert statement proposal intriguing, but also too vague to have a very strong opinion on at this point.  I.e., would it still support the ability to associate a user-facing message with the assert (which I've appreciated having), how would it help support the various flavors of assert, etc.

One other related issue that we didn't really get to today is whether we have any hope of unifying compile-time assertions with these two flavors of assert (and warning, error, and possibly note), or whether they'll forever have to be a completely different thing.

**Comment by mppf** [comment]
@bradcray - I think that going with `if checksAreOn then assert(...)` for now would be reasonable (and this approach had the most support in the discussion). I think there are lots of ways that we could extend this or add to it without making breaking changes. However it is unclear to me from your previous comment if you are saying we need to do something else before we go with that strategy. E.g. do you think we need to do more investigation of the alternatives you have mentioned?

**Comment by lydia-duncan** [comment]
Since this is additive, I think are considering this post-2.0 and removing the tag.

**Comment by mppf** [comment]
Noting something @twesterhout brought up in Gitter. He ran into performance challenges when using `assert` that could be resolved if the compiler were to rewrite `assert(condition, f())` as something like `if !condition then halt(f())`. This is slightly different from the ideas brought up so far in this issue.
